### PR TITLE
Donate Manager implementation.

### PR DESCRIPTION
**Pull request by geoffl**
I've put everything into a donate manager singleton to group things together even though it has stuff both in the backend and frontend, but it's not too bad.  In debug powertoys there's one or two things that can be touched by both the backend and frontend, but I'm just not going to worry about it for now.

### CODE REVIEW

File: tv/lib/donate.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+from miro import signals
+
+from miro.frontends.widgets import donate
+
+from miro.plat.frontends.widgets.threads import call_on_ui_thread
+
+class DonateManager(object):
+    NOTHANKS_MULTIPLIER = [50, 100]
+    URL_PATTERN = 'http://getmiro.com/give/m%s/'
+    PAYMENT_URL = 'https://www.stayclassy.org/checkout/donation?cid=7214'
```
Comment (paulswartz) [inline tv/lib/donate.py]: Perhaps these should be in appconfig?

File: tv/lib/donate.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+from miro import prefs
+from miro import signals
+
+from miro.frontends.widgets import donate
+
+from miro.plat.frontends.widgets.threads import call_on_ui_thread
+
+class DonateManager(object):
+    NOTHANKS_MULTIPLIER = [50, 100]
+    URL_PATTERN = 'http://getmiro.com/give/m%s/'
```
Comment (paulswartz) [inline tv/lib/donate.py]: Also, this URL pattern doesn't appear to work.

File: tv/lib/donate.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    def reset(self):
+        for pref in [prefs.DONATE_NOTHANKS, prefs.LAST_DONATE_TIME,
+                     prefs.DONATE_COUNTER]:
+            app.config.set(pref, pref.default)
+
+    def show_donate(self, url=None, payment_url=None):
+        if not url:
+            args = [1, 2, 3]
+            try:
+                url = self.URL_PATTERN % args[self.donate_nothanks]
```
Comment (paulswartz) [inline tv/lib/donate.py]: Is this just `URL_PATTERN % min(3, self.donate_nothanks + 1)` or am I missing something?

File: tv/lib/donate.py (hunk 1 above)
Comment (geoffl) [inline tv/lib/donate.py]: perhaps, but these are read-only anyway so I didn't put it in.  We can do it like the app serial though but not sure if it's worth it.  Either way I don't have strong attachment.  Do you prefer these to be in app config?

File: tv/lib/donate.py (hunk 3 above)
Comment (geoffl) [inline tv/lib/donate.py]: no, that's basically it.  There used to be something which handled another case but is no longer there.  Anyway, try: except: may not be the best way to express it but none of this is very performance critical so it doesn't seem too bad.

Also, what did you mean above by the URL not working?  The web pages are not in place so they won't navigate to the right place.  I think you get redirected to the main page.

File: tv/lib/donate.py (hunk 1 above)
Comment (paulswartz) [inline tv/lib/donate.py]: That's the way we do other "read-only" items; they're at least overridable by the app.config file.

File: tv/lib/donate.py (hunk 3 above)
Comment (paulswartz) [inline tv/lib/donate.py]: Okay.  Is there a way to disable this while the donation pages aren't available? It'll just be confusing if users get a popup of the Miro homepage.

File: tv/lib/prefs.py [hunk 4]
```diff
@@ -62,6 +62,14 @@ def __ne__(self, other):
         return self.key != other.key
 
 # These are normal user preferences.
+DONATE_PAYMENT_URL          = Pref(key='donatePaymentURL',      default='https://www.stayclassy.org/checkout/donation?cid=7214', platformSpecific=False)
+DONATE_URL_TEMPLATE         = Pref(key='donateURLTemplate',     default='http://getmiro.com/give/m%s/', platformSpecific=False)
+DONATE_ASK1                 = Pref(key='donateAsk1',            default=10,    platformSpecific=False)
```
Comment (paulswartz) [inline tv/lib/prefs.py:67]: Can you document what these variables are?

Comment (geoffl) [inline tv/lib/prefs.py:67]: added better annotation in the lib/donate.py file about what these do.

Also fixed an apparent bug where seemingly the donate window could call show() in the backend.

**Pull request merged**